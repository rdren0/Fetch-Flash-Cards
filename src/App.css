.App {
  text-align: center;
}

.App-link {
  color: #61dafb;
}

@keyframes App-logo-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

body{
  margin: 0px;
  background-color: #2E6171;
  text-align: center;
}
h1{

}

.header{
margin: 0px;

}

.button-holder{
  background-color: #556F7A;
  height: 60px;
}

.container{
  background-color: #2E6171;
}