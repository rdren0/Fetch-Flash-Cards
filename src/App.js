import React, { Component } from 'react';
import Header from './Components/Header.js'
import Container from './Components/Container.js';
import data from './data.js';
import loading from './Images/loading.svg';


export class App extends Component {
  constructor() {
    super();
    this.state = {
      questions: [],
      completed: [],
      selection: [],
      loading: true
    }
  }


  showAll = () =>{
    let allCards = this.state.questions.filter(card =>{
      return (!this.state.completed.includes(card.ID))
    }).sort(() => Math.random() - 0.5);
    this.setState({
      selection: allCards
    });    
  }


  sortCards = (e) =>{
    let value = e.target.innerText;
    let sortedCards =  this.state.questions.filter(card =>{
      return card.Difficulty.toString() === value;
    })
    this.setState({
        selection: sortedCards
    }); 
  }

  onlyCompleted = () =>{
    let completedCards = this.state.questions.filter(card =>{
      return (this.state.completed.includes(card.ID))
    })
    this.setState({
        selection: completedCards
    }); 
  }


  addCompleted = (ID, status) => {
    let newState;
    if (this.state.completed.includes(ID) && status === true) {
      newState = this.state.completed.filter(card => card !== ID);
    } else{
      newState = [...this.state.completed, ID];
    this.setState({
      completed : newState
    }, () => {
      this.saveToStorage();
    });
    }
  }


  saveToStorage = () => {
    localStorage.setItem('Completed', JSON.stringify(this.state.completed));
  }


  componentDidMount() {
    fetch("https://fe-apps.herokuapp.com/api/v1/memoize/1901/rachael-fetch/flashcards")
    .then(response => response.json())
    .then(questions => this.setState({questions: questions.flashCards}))
    .catch(err => {
      throw new Error(err);
    })
    const completedCards = JSON.parse(localStorage.getItem('Completed')) || [];
      this.setState({
        completed: completedCards,
        loading: false
      });
  }
  render() {
     if(this.state.loading){
      return(
        <div>
        <img className="App-loading" src={loading} alt="loading-icon" />
          <p>
            Loading flashCards...          
          </p>
        </div>)
    }else{
      return(
        <div>
          <Header/>
          <div className="button-holder">
            <button id= "showAll" onClick={this.showAll}>All</button>
            <button id="easy"onClick={this.sortCards}>Easy</button>
            <button id="medium" onClick={this.sortCards}>Medium</button>
            <button id="hard" onClick={this.sortCards}>Hard</button>
            <button className= "completed-button" onClick={this.onlyCompleted}>Completed</button>
          </div>
          <hr/>
          <Container 
          data = {this.state.selection}
          completed = {this.state.completed}
          addCompleted = {this.addCompleted}
          />
      </div>
      );
    }
  }

}

export default App;
