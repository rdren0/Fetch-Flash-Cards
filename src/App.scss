.App {
  text-align: center;
}

.App-link {
  color: #61dafb;
}

.App-loading{
  animation: App-loading-spin infinite 10s linear;
  height: 40vmin;
  pointer-events: none;
  margin-top: 50px;

}

@keyframes App-loading-spin {
  from {
    transform: rotate(0deg);
  }
  to {
    transform: rotate(360deg);
  }
}

hr{
  border: 1.5px solid black;
  margin-bottom: 0px;
  margin-top: 20px;
}

.button-holder{
  display: inline-block;
  text-align: center;
  border-radius: 15px;
  margin-right: auto;
  margin-left: auto;
  background-color: #556F7A;
  /*height: 60px;*/
  margin-bottom: 10px;
}

button{
  height: 45px;
  width: 90px;
  border: 2px solid black;
  border-radius: 15px;
  margin: 8px;
  background-color: #DD7373;
  color: black;
}

button:hover{
  background-color: #E5C3D1;
  cursor: pointer;
}

.completed-button{
 background-color: #87B38D
}

.completed-button:hover{
  background-color: #23ae87;
}

.bulb{
  height: 150px;
}
