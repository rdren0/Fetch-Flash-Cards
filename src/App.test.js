import React from 'react';
import ReactDOM from 'react-dom';
import App from './App';
import { shallow } from 'enzyme';
import { configure } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({ adapter: new Adapter() });

const setStorage = jest.spyOn(Storage.prototype, 'setItem');
const getStorage = jest.spyOn(Storage.prototype, 'getItem');

const testEasy =  {target: { innerText: 'easy' }} 
const testCompleted = [8,9]


const mockCards = [{
  "ID": 7,
  "Question": "What should always be the last function on a fetch call?",
  "Answer": ".catch()",
  "Difficulty": "Easy"
}, {
  "ID": 8,
  "Question": "What method should our fetch call always live in?",
  "Answer": "componentDidMount()",
  "Difficulty": "Easy",
  "status": true
}, {
  "ID": 9,
  "Question": "A ___ is a computer that hosts data so that one can remotely access this data.",
  "Answer": "Server",
  "Difficulty": "Easy"
}]


describe('App', () => {
  let wrapper;
  
  beforeEach(() =>{
    wrapper = shallow(
      <App
      />
    )
  })
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<App />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('should match the snapshot with all data passed in correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should match the snapshot when loading data', () => {
    wrapper.setState({ loading: true });
    expect(wrapper).toMatchSnapshot();
  });

  it('should show all cards expected completed ones', () =>{
      wrapper.setState({ questions: mockCards });
      wrapper.setState({ completed: [8,9] });
      wrapper.find('#showAll').simulate('click');
      expect(wrapper.state("selection")).toEqual([{
        "ID": 7,
        "Question": "What should always be the last function on a fetch call?",
        "Answer": ".catch()",
        "Difficulty": "Easy"
      }])
  });

  it('show sort Cards by difficulty', () => {
    wrapper.setState({ questions: mockCards });
    wrapper.find('#medium').simulate('click', testEasy);
    expect(wrapper.state("selection")).toEqual([])
  });

  it('show only display completed Cards', () => {
    wrapper.setState({ questions: mockCards });
    wrapper.setState({ completed: testCompleted });
    wrapper.find('.completed-button').simulate('click', testEasy);
    expect(wrapper.state("selection")).toEqual([mockCards[1], mockCards[2]])
  });

  it('should add card to completed', () => {
    wrapper.setState({ completed: testCompleted });
    wrapper.instance().addCompleted(8, true);
    expect(wrapper.state("completed")).toEqual([8,9])

  });

  it('should add card to completed', () => {
    wrapper.instance().saveToStorage();
    expect(setStorage).toHaveBeenCalled();
  });

  it.skip('should fetch data', () =>{
    expect(componentDidMount).toHaveBeenCalled();
  })

})
