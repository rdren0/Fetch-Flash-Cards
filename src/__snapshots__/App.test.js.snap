// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot when loading data 1`] = `
<div>
  <img
    alt="loading-icon"
    className="App-loading"
    src="loading.svg"
  />
  <p>
    Loading flashCards...
  </p>
</div>
`;

exports[`App should match the snapshot with all data passed in correctly 1`] = `
<div>
  <Header />
  <div
    className="button-holder"
  >
    <button
      id="showAll"
      onClick={[Function]}
    >
      All
    </button>
    <button
      id="easy"
      onClick={[Function]}
    >
      Easy
    </button>
    <button
      id="medium"
      onClick={[Function]}
    >
      Medium
    </button>
    <button
      id="hard"
      onClick={[Function]}
    >
      Hard
    </button>
    <button
      className="completed-button"
      onClick={[Function]}
    >
      Completed
    </button>
  </div>
  <hr />
  <Container
    addCompleted={[Function]}
    completed={Array []}
    data={Array []}
  />
</div>
`;
